Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,396 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: S. Gangloff
+-/
+
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.Int.Basic
+import Mathlib.Algebra.Module.Basic
+import Mathlib.Topology.Basic
+import Mathlib.Topology.UniformSpace.Pi
+import Mathlib.Topology.UniformSpace.Basic
+import Mathlib.Topology.Sets.Opens
+import Mathlib.Topology.Instances.Discrete
+import Mathlib.Algebra.Group.Basic
+import Mathlib.Data.Int.Interval
+import Mathlib.Data.Real.Basic
+import Mathlib.Order.Filter.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.Basic
+import Mathlib.Topology.Instances.EReal.Lemmas
+import Mathlib.Data.Finset.Basic
+import Mathlib.Logic.Equiv.Defs
+
+/-!
+# Symbolic dynamics over `ℤ^d`
+
+We set up a minimal API for the full shift over a finite discrete alphabet, cylinders,
+patterns, and subshifts defined by forbidden patterns. We also define a language size on
+boxes and a (limsup-based) notion of topological entropy.
+
+## Main definitions
+
+* `Zd d` : the lattice `ℤ^d` as functions `Fin d → ℤ`.
+* `FullShiftZd A d` : configurations `Zd d → A` with the product topology.
+* `FullShiftZd.shift` : the `ℤ^d`-action by translation.
+* `cylinder U x` : cylinder set fixing the coordinates in `U ⊆ Zd d` to those of `x`.
+* `Pattern A d` : finite patterns (support + values).
+* `Pattern.occursIn` : occurrence of a pattern at a position.
+* `forbids F` : configurations avoiding every pattern in `F`.
+* `Subshift` : closed, shift‑invariant subsets of the full shift.
+* `box n` : the cube `[-n,n]^d` as a finset of `Zd d`.
+* `languageCard` / `patternCount` : number of patterns seen on a box.
+* `limsupAtTop` : `limsup` of a real sequence along `atTop` (infimum of eventual upper bounds).
+* `entropy` : limsup language growth per unit volume.
+
+## Notation/Conventions
+
+* The alphabet `A` is assumed finite, discrete, and inhabited throughout. The product topology
+  on configurations is used.
+* Cylinders are defined by equality on finitely many coordinates, hence are clopen.
+
+-/
+
+open Set Topology
+open Filter
+
+namespace SymbolicDynamics
+
+
+variable {A : Type*} [Fintype A] [DecidableEq A] [Inhabited A]
+  [TopologicalSpace A] [DiscreteTopology A]
+
+variable {d : ℕ}
+
+/-- The lattice `ℤ^d` as functions `Fin d → ℤ`. -/
+def Zd (d : ℕ) := Fin d → ℤ
+
+/-- Decidable equality on `Zd d`. -/
+@[instance]
+def Zd.decidableEq (d : ℕ) : DecidableEq (Zd d) :=
+  inferInstanceAs (DecidableEq (Fin d → ℤ))
+
+/-- Pointwise addition on `ℤ^d`. -/
+instance : Add (Zd d) where
+  add := fun u v i ↦ u i + v i
+
+/-- `ℤ^d` is an additive commutative group (pointwise). -/
+instance : AddCommGroup (Zd d) := Pi.addCommGroup
+
+/-! ## Full shift -/
+
+/-- The full shift over `ℤ^d` with alphabet `A`. -/
+@[reducible]
+def FullShiftZd (A : Type*) (d : ℕ) := Zd d → A
+
+/-- Product topology on the full shift. -/
+instance : TopologicalSpace (FullShiftZd A d) := Pi.topologicalSpace
```

Comment:
Because you used `reducible` above, you don't get to choose this, it's forced upon you automatically. I'd suggest just deleting this instance, and same for the one below